File: ios/stereoCamera/stereoCamera/Comm/Commands/SetOverlay.swift
//
//  SetOverlay.swift
//  stereoCamera
//
//  Created by hallmarklabs on 6/27/18.
//  Copyright © 2018 cody. All rights reserved.
//

import Foundation

class SetOverlay : Command
{
    var overlay = Overlay.NONE

    override init()
    {
        super.init()
        doInit()
    }
    
    init(_ overlay: Overlay)
    {
        super.init()
        self.overlay = overlay
        doInit()
    }
    
    private func doInit()
    {
        cmdtype = CommandTypes.SET_OVERLAY
        expectsResponse = true
    }
    
    override func send(comm: Comm) -> Bool
    {
        let result = super.send(comm: comm)
        if (result == false) { return false }
        
        var bytes:[UInt8] = []
        bytes += Bytes.toByteArray(overlay)
        
        let sz = comm.write(buf: bytes)
        
        return (sz <= 0) ? false : true
    }
    
    override func receive(comm: Comm) -> Bool
    {
        let result = super.receive(comm: comm)
        if (result == false) { return false }
        
        let (buf, sz) = comm.read(sz: 1)
        if (sz <= 0) { return false }
        
        overlay = Bytes.fromByteArray(buf)
        return true
    }
}

File: android/app/src/main/java/com/munger/stereocamera/ip/command/commands/SetOverlay.java
package com.munger.stereocamera.ip.command.commands;

import com.munger.stereocamera.ip.command.Comm;
import com.munger.stereocamera.ip.command.Command;
import com.munger.stereocamera.widget.PreviewOverlayWidget;

import java.io.IOException;

public class SetOverlay extends Command
{
	public PreviewOverlayWidget.Type type;

	public SetOverlay()
	{
		super();
		doInit();
	}

	public SetOverlay(PreviewOverlayWidget.Type type)
	{
		super();
		this.type = type;
		doInit();
	}

	private void doInit()
	{
		cmdtype = Type.SET_OVERLAY;
		expectsResponse = true;
	}

	@Override
	public boolean send(Comm comm)
	{
		boolean result = super.send(comm);
		if (!result)
			{ return false; }

		try
		{
			comm.putByte((byte) type.ordinal());
		}
		catch(IOException e){
			return false;
		}

		return true;
	}

	@Override
	public boolean receive(Comm comm)
	{
		boolean result = super.receive(comm);
		if (!result)
			{ return false; }

		try
		{
			byte b = comm.getByte();

			PreviewOverlayWidget.Type[] types = PreviewOverlayWidget.Type.values();
			if (b < 0 || b >= types.length)
				return false;

			type = types[b];
		}
		catch(IOException e){
			return false;
		}

		return true;
	}
}
